### ISSUE TITLE
Support for FIDEDUPERANGE

### ISSUE THREAD

**Issue by luke-jr** [openzfs/zfs#11065]
Linux 5.4 added an ioctl to inform filesystems of duplicate data, for deduplication. It would be nice if ZFS supported it.

Use case: a cron job could regularly find data to deduplicate without the memory overhead of current ZFS deduplication.

**Comment by Ornias1993** [comment]
Are you sure that said function is compatible with ZFS?
Not part of LVM or something?

**Comment by Rydako** [comment]
https://man7.org/linux/man-pages/man2/ioctl_fideduperange.2.html

"This ioctl operation first appeared in Linux 4.5.  It was previously known as BTRFS_IOC_FILE_EXTENT_SAME and was private to Btrfs."

Isn't LVM specific, and is already used by BTRFS it seems.

**Comment by luke-jr** [comment]
>Are you sure that said function is compatible with ZFS?

At least in some contexts, it should be? ZFS already deduplicates snapshots, for example.

I'm not familiar with the more generalised existing ZFS deduplication, but I don't know why it wouldn't prove this function is possible.

>"Both files must reside within the same filesystem" I'm by no means an expert on the inner workings of ZFS, but wouldn't this caveat mean this could only be used to dedup within a single dataset?

If you add such a limitation, it might be nice to have an option to not enforce it. :)

**Comment by shodanshok** [comment]
This basically is a rehash of [`reflink` support](https://github.com/openzfs/zfs/issues/405). A much needed feature, I must say...

**Comment by strugee** [comment]
Note that this issue is theoretically unblocked since reflink support has landed.

**Comment by iakat** [comment]
Related: https://github.com/Mic92/zfs-dedup https://github.com/Mic92/zfs/tree/fideduperange
